README.md

```diff
@@ -122,6 +122,239 @@ the components involved in the test:
 cd test/integration && dlv test -- -test.run ^TestCrud$
 ``
 
+#### E2E
+
+The Federation-v2 E2E tests can run in an *unmanaged*, *managed*, or *hybrid*
+modes. For both unmanaged and hybrid modes, you will need to bring your own
+clusters. The managed mode runs similarly to integration as it uses the same
+test fixture setup. All of these modes run CRUD operations. CRUD here means
+that the tests will run through each of the requested federated types and
+verify that:
+
+1. the objects are created in the target clusters.
+1. an annotation update is reflected in the objects stored in the target
+   clusters.
+1. a placement update for the object is reflected in the target clusters.
+1. deleted resources are removed from the target clusters.
+
+The read operation is implicit.
+
+##### Managed
+
+The E2E managed tests will spin up the same environment as the
+[Integration](README.md#integration) tests described above and run [CRUD
+(create-read-update-delete)
+checks](https://github.com/kubernetes-sigs/federation-v2/blob/master/test/e2e/crud.go) for
+federated types against that federation. To run:
+
+ - ensure the same binaries are available as described in the
+   [Integration](README.md#integration) section.
+
+To run tests for all types:
+
+```bash
+cd test/e2e
+go test -v
+```
+
+To run tests for a single type:
+
+```bash
+cd test/e2e
+go test -args -v=4 -test.v --ginkgo.focus='"FederatedSecret"'
+```
+
+It may be helpful to use the [delve
+debugger](https://github.com/derekparker/delve) to gain insight into
+the components involved in the test:
+
+```bash
+cd test/e2e
+dlv test -- -v=4 -test.v --ginkgo.focus='"FederatedSecret"'
+```
+
+##### Unmanaged and Hybrid Cluster Setup
+
+The difference between unmanaged and hybrid is that with hybrid, you run
+the federation-v2 controllers in-process as part of executing the test. This
+helps in running a debugger to debug anything in the controllers. On the other
+hand with unmanaged, the controllers are already running in the K8s cluster.
+Both methods require the clusters to have already been joined.
+
+###### Create Clusters
+
+The quickest way to set up clusters for use with unmanaged and hybrid modes is
+to use [minikube](https://kubernetes.io/docs/getting-started-guides/minikube/).
+Once you have minikube installed run:
+
+```bash
+minikube start -p clusterA
+```
+
+###### Deploy the Cluster Registry
+
+Get the latest version of the [Cluster
+Registry](https://github.com/kubernetes/cluster-registry/releases) and run the
+following:
+
+```bash
+crinit aggregated init mycr --host-cluster-context=clusterA
+```
+
+You can also specify your own cluster registry image using the `--image` flag.
+
+###### Deploy Federation
+
+First you'll need to create the namespace to deploy the federation into:
+
+```bash
+kubectl create ns federation
+```
+
+Then run `apiserver-boot` to deploy. This will build the federated API server
```

Comment:
Consider documenting the requirement to configure login to the docker hub:

https://github.com/kubernetes-sigs/federation-v2/pull/57/files#diff-1ed12d3b5689c7a45b1c21e93e23e3d3R40

Also, have you tested the deployed controller manager?  I needed to create a permissive role binding or the controller manager wouldn't be able to perform cluster list/watch and the apiserver reported errors related to webhook configuration:

https://github.com/kubernetes-sigs/federation-v2/pull/57/files#diff-1ed12d3b5689c7a45b1c21e93e23e3d3R64

Reply:
I don't necessarily want to be prescriptive to docker hub e.g. I use google's container registry. But I think a generic "login" comment perhaps with docker as an example is useful.

Yes, I have tested the deployed controller manager in `minikube` version `v0.25.2`, but did not run into any permission issues. Interestingly enough though, during my testing of this deployment in a public cloud environment on Friday, both the `apiserver` and `controller` fail due to a lack of various RBAC permissions that were not encountered with `minikube`. Basically, `apiserver-boot` does not generate a deployment config that includes all the necessary RBAC objects for an aggregated API server.

I think that once we have an officially released alpha version of federation-v2 with a container image, we should commit the `apiserver.yaml` config with all the necessary pieces: pointing to the released container image, the memory limit update, and necessary RBAC objects. I think there is even benefit to committing it now with the expectation that users update the container image they are using.

Reply:
Done.